### PR TITLE
feat: prototype to use checkboxes in the extension interface

### LINKED ISSUE #6: Add 'deprecated in v5' classes file
This is a great extension - nice job on building this :-)

Not too important right now really, but could be nice to add a deprecated-in-v5 JSON file to track classes which are deprecated such as:

- `.carousel-dark` (https://getbootstrap.com/docs/5.3/components/carousel/#dark-variant)
- `.navbar-dark` (https://getbootstrap.com/docs/5.3/components/navbar/#color-schemes)
- `.dropdown-menu-dark` (https://getbootstrap.com/docs/5.3/components/dropdowns/)

### PR DESCRIPTION
**Pull request by julien-deramond**
Closes #6

## Description

This PR replaces buttons with checkboxes in the interface in order to avoid the exclusive choice of highlighting deprecated classes in BS3, and BS4. The reset button stays there as it can ease the manipulation to reset all the checkboxes with one click.

This PR also introduces a new feature where Bootstrap v5 deprecated classes are highlighted in yellow. We're talking here about the classes that can still work in Bootstrap v5, but will be deleted in v6. So having them highlighted can show in a website what could be tackled right away before the v6, but which is not mandatory.

The full list of deprecated Bootstrap v5 classes is:
- `.btn-close-white`
- `.carousel-dark`
- `.dropdown-menu-dark`
- `.navbar-dark`
- `.navbar-light`
- `.text-black-50`
- `.text-muted`
- `.text-white-50`

**This list can't be retrieved automatically!** That's why `script/build.js` now has a special treatment where it will create "manually" the `./data/bootstrap-v5-deprecated-classes.json` (with Bootstrap 5 deprecated classes list directly in the script). However, the key will be different: `deprecated` instead of `removed`. Thanks to that, we'll be able to automatically render the elements with a different color; here, yellow.

After being merged, the browser extension will need to be released as a 0.2.0 version, and the screenshots and description might need to be updated too.

### Rendering (screenshots)

Screenshot 2024-10-02 at 17 26 45
Screenshot 2024-10-02 at 17 25 56

---

## Remaining Sub-tasks

- [x] Try out checkboxes in the interface
- [x] Think about having only checkboxes, checkboxes only if there's a Bootstrap 5 use case (see https://github.com/julien-deramond/bootstrap-deprecated-classes-extension/issues/6)
- [x] Handle the reset button
- [x] Add coliff to the co-authors 
- [x] Refine the description here based on what's really done (see https://github.com/julien-deramond/bootstrap-deprecated-classes-extension/pull/8#issuecomment-2388988970 for instance)
- [x] Check all deprecated classes in BS 5
- [x] Check that it works for all browsers
- [x] Check why the state seems to be lost when we close the extension panel and reopen it

### CODE REVIEW

**Comment by coliff** [comment]
Just built this locally and tested it out in Edge - works great. I like the idea of being able to toggle them with checkboxes.

Just discovered another checkbox that could be added... right now negative margin classes (e.g. `mb-n1` are highlighted as deprecated, but even though they aren't included in BS5 CDN, they are available at built time. I use them a lot so it'd be good to have option to not have these highlighted). We should of course make sure that `ml-n1` are still highlighted (as they should be `ms-n1` in BS5).

**Comment by julien-deramond** [comment]
I like the idea. Essentially, this would mean that the default "Bootstrap 5" (or another version) checkbox would represent Bootstrap's standard configuration, as found in the CDN. Then, there could be an "Advanced mode" option that lets users fine-tune Bootstrap's content based on the available customization options at [Bootstrap's customization page](https://getbootstrap.com/docs/5.3/customize/options/).

For instance, if you click on "Advanced mode," additional checkboxes could appear for all `$enable-*` options, allowing you to toggle features. In your specific case, the `$enable-negative-margins` option would be unchecked by default, and you could enable it to remove the related deprecations in the interface.

Alternatively, we could group all classes related to `$enable-*` options under a category called "Optional classes." By default, these would be highlighted in a different color, and users could uncheck a box labeled "Optional classes" to toggle them. This would apply to all Bootstrap versions, making it a more generic solution, likely simplifying the form interface for the extension.

This approach could be implemented in a separate PR as a new feature.

**Comment by julien-deramond** [comment]
I've made several modifications in this PR to experiment with new ideas. Specifically, the interface now includes a checkbox designed to manage deprecated Bootstrap 5 classes. In this particular use case, these deprecated classes would be rendered with a yellow color for distinction.

The intention behind the checkbox is to highlight that this is a unique scenario, though the explanation may need refinement.

Below are some examples:

Screenshot 2024-10-02 at 17 26 45
Screenshot 2024-10-02 at 17 25 56

The reset button is now functional (tested only on Firefox for now).

Under the hood, a new `data/bootstrap-5-deprecated-classes.json` is manually "generated", based on specific classes defined in `script/build.mjs` like this:

```js
// Generate ./data/bootstrap-${version}-deprecated-classes.json` manually for the latest version
const lastestVersionOutputFilePath = `./data/bootstrap-${maximumBootstrapVersion}-deprecated-classes.json`
await fs.writeFile(lastestVersionOutputFilePath, JSON.stringify({
  deprecated: [
    ".carousel-dark",
    ".navbar-dark",
    ".dropdown-menu-dark"
  ] 
}))
```

The added JS part is a bit ugly but seems functional. Refactoring could be done in a follow-up PR.

@coliff, if you have some time, I'd really appreciate your feedback as a potential user.

Just a note: the feature mentioned in https://github.com/julien-deramond/bootstrap-deprecated-classes-extension/pull/8#issuecomment-2387527116 will be handled in a separate PR. I plan to create a dedicated issue for it soon. :)

**Comment by coliff** [comment]
Nice! Had a quick test of this on Edge (Windows 11) and reset button is working well and the new deprecated in v5 check is working (tested on https://getbootstrap.com/docs/5.3/components/carousel/#dark-variant).

I noticed that if you check a checkbox and then click away to dismiss the popup and then display it again the checkboxes are unchecked (despite it actually being active). (hope that makes sense!). No big deal really though...

File: script/build.mjs [hunk 1]
```diff
@@ -32,18 +33,28 @@ for (const version of bootstrapVersions.slice(0, -1)) {
   await fs.writeFile(outputFilePath, JSON.stringify(outputJSONContent))
 }
 
+// Generate ./data/bootstrap-${version}-deprecated-classes.json` manually for the latest version
+const lastestVersionOutputFilePath = `./data/bootstrap-${maximumBootstrapVersion}-deprecated-classes.json`
```
Comment (coliff) [inline script/build.mjs:37]: typo: think `lastestVersionOutputFilePath` should be `latestVersionOutputFilePath`

File: script/build.mjs [hunk 2]
```diff
@@ -32,18 +33,28 @@ for (const version of bootstrapVersions.slice(0, -1)) {
...
+await fs.writeFile(lastestVersionOutputFilePath, JSON.stringify({
+  deprecated: [
+    ".carousel-dark",
```
Comment (coliff) [inline script/build.mjs:40]: could add `btn-close-white` class here.
REF: https://getbootstrap.com/docs/5.3/components/close-button/#dark-variant

Comment (julien-deramond) [reply script/build.mjs:41]: Thanks, added with https://github.com/julien-deramond/bootstrap-deprecated-classes-extension/pull/8/commits/5ba79acac0342e965bf8890c5c860e684c79bd14
Still need to check the entire Sass code of Bootstrap to find the deprecated variables. There are also some utilities to add, if I remember well.

**Comment by coliff** [comment]
yeah - works great! 

🚢 Ship it!

**Pull request merged**